### PR TITLE
Improve debugging experience

### PR DESCRIPTION
**Pull request by sharwell**
* Do not optimize test framework (stepping in debugger will have full information about parameters, locals, and breakpoints)
* Embed symbols in assemblies (eliminates the need to copy PDB files)
* Embed sources in assemblies (eliminates the need to download sources separately)

### CODE REVIEW

**Comment by jnm2** [comment]
>     * Do not optimize test framework (stepping in debugger will have full information about parameters, locals, and breakpoints)

I'm concerned about this. Can you give us some idea of what performance we are giving up as well as some idea of how much time this would save you and how often this would happen? Stepping through release builds of libraries is a well-known scenario. I'm not sure what makes NUnit a special case. 

>     * Embed symbols in assemblies (eliminates the need to copy PDB files)

I'm a fan. Isn't .snupkg the current recommendation and also working out of the box these days for anyone with VS2019+, with a very simple fallback (add symbol download address) for folks on older VS?

>     * Embed sources in assemblies (eliminates the need to download sources separately)

I'm a fan. I do this absolutely everywhere I'm allowed.

**Comment by rprouse** [comment]
@sharwell because of our history, I trust you and this looks good, but I would appreciate it if you could provide a bit more explanation or point us to the documentation for the build properties you introduced. I like the change, but I'd like to read up a bit more on it if possible.

**Comment by sharwell** [comment]
> I'm concerned about this. Can you give us some idea of what performance we are giving up as well as some idea of how much time this would save you and how often this would happen?

Since NUnit doesn't have heavy computation loops, I'm not expecting a significant observable change. The most likely observations will be increases in allocations associated with methods marked `async` in the NUnit code base, though these should not be significant compared to the cost of the tests themselves.

The best test case would likely be a project with a very large number of very small tests.

**Comment by sharwell** [comment]
I'm going to rebase this change to address the merge conflict.

**Comment by jnm2** [comment]
@sharwell I'm still curious what makes NUnit special, or is this coming from a blanket recommendation that all .NET libraries should consider switching to shipping debug builds?

**Comment by sharwell** [comment]
More the idea that unit testing frameworks are part of the developer's inner loop more than part of production code. The case where I found myself wishing for better debugging was trying to step through `ThrowsAsync` to see how it interacted with synchronization contexts.

**Comment by jnm2** [comment]
That helps, thanks! This type of change would thus not be applicable to the NUnit console, engine, or VSTest adapters then, I'm thinking?

File: build.cake [hunk 1]
```diff
@@ -325,21 +318,9 @@ Task("CreateImage")
...
-    .IsDependentOn("CreateImage")
-    .Description("Source-indexes PDBs in the images directory to the current commit")
-    .Does(() =>
-    {
-        var settings = new GitLink3Settings
-        {
-            BaseDir = PROJECT_DIR
-        };
-        GitLink3(GetFiles($"{CurrentImageDir}**/*.pdb"), settings);
-    });
```
[the remark this answers is not in the document]
Comment (jnm2) [reply build.cake]: When I [did the numbers](https://github.com/nunit/nunit/issues/2421#issuecomment-358849839), the portable PBD+embedded source was 3.75× the size and Windows PDB+sourcelink was 2.51× the size. I don't think size should be the determining factor though.

Embedded source requires VS 2017 Update 5+ for the debugger to locate the source. Do you know what version of VS is required to locate the source when the PDB is also embedded in the assembly?

I'm thinking that's far enough behind us now that it isn't a concern. Guessing that latest 2017 and 2019 both work.

**Comment by jnm2** [comment]
https://github.com/nunit/nunit/blob/master/README.md#debugger-source-stepping also needs to be updated or deleted.

**Comment by clairernovotny** [comment]
Just came across this and it looks like GitLink was generating invalid Source Link data:

image

It's embedding its data as lower-case instead of matching what's in the PDB so we can't correctly substitute.  Moot point now but something we should be aware of.

**Comment by mikkelbu** [comment]
@sharwell, @jnm2 After this change is seems that the debugger will always enter NUnit code when people are debugging their tests (no matter if "Enable Just My Code" is checked or not). It has reported in https://groups.google.com/g/nunit-discuss/c/f2yV4G_kdi8/m/ypWZfLSEDgAJ and I can also reproduce this. Do you know if it is possible to configure VS such that it does not enter NUnit code.

**Comment by jnm2** [comment]
Even with Just My Code disabled?

**Comment by jnm2** [comment]
I see, even with Just My Code turned off. I'm in favor of moving to the snupkg format then, since that uses the well-known nuget.org symbol server that has been built into VS for a while.

**Comment by jnm2** [comment]
https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/nuget#symbol-packages
https://docs.microsoft.com/en-us/nuget/create-packages/symbol-packages-snupkg
https://devblogs.microsoft.com/nuget/improved-package-debugging-experience-with-the-nuget-org-symbol-server/

**Comment by sharwell** [comment]
@mikkelbu I would suggest filing an issue through [Developer Community](https://developercommunity.visualstudio.com/). I can't wrap my head around the idea that someone could have Just My Code enabled and not think it's a bug, so the response that makes the most sense to me is "obviously this behavior is correct, so I'm not sure what the question is 😕 ".

**Comment by jnm2** [comment]
@sharwell Just My Code is a lifesaver when you are working with libraries that handle exceptions or when you don't want extra waits every time you debug where VS loads symbols (even when using a whitelist).

**Comment by mikkelbu** [comment]
I've created the following issue in Developer Community - https://developercommunity2.visualstudio.com/t/The-debugger-enters-code-in-PackageRefer/1321266

**Comment by gfoidl** [comment]
From [.NET Just My Code](https://docs.microsoft.com/en-us/visualstudio/debugger/just-my-code?view=vs-2019#BKMK__NET_Framework_Just_My_Code):

> In .NET projects, Just My Code uses symbol (.pdb) files and program optimizations to classify user and non-user code. The .NET debugger considers optimized binaries and non-loaded .pdb files to be non-user code.

So one can argue that the .NET Debugger is correct by thinking NUnit* is user-code, as pdf is there and program optimizations are off.

But I agree that this is counterintuitive and "Just my code" should be tightened to only consider "my" code.

\* Moq, etc. which also embeds the pdb have the same "issue".

**Pull request merged**